test/support/sequence_coverage_reporting.rb

```diff
@@ -60,49 +65,141 @@ def add_assertion_location(location)
+  end
+
+  def method_name
+    @method_name ||= backtrace[sequence_call_index].to_s.match(/`(.*)'/)&.[](1)
+  end
+
+  # This adjusts sequence_call_index so that when assertions are wrapped in a
+  # warning block, this index refers to the method call inside the warning block
+  # rather than the warning call itself
+  def adjust_index_to_handle_blocks
+    return if sequence_call_index.blank? || assertion_call_index.blank?
+
+    (assertion_call_index...sequence_call_index).to_a.reverse_each do |index|
+      self.sequence_call_index = index if backtrace[index].to_s.match?(/block .*in #{method_name}/)
+    end
+  end
+
+  def location
+    return unless sequence_call_index.present? && assertion_call_index.present?
+
+    [backtrace[sequence_call_index], backtrace[assertion_call_index + 1]]
+      .uniq
+      .map { |location| location&.to_s&.match(LINE_REGEX)&.[](1) }
+      .join(CALL_SEPARATOR)
+  end
+end
+
+# This class generates a csv file with information on how many times each
+# assertion passed or failed in unit tests. The output differentiates direct and
+# indirect assertion calls. Direct assertion calls are assertions that are
+# called directly in a test or method being tested. Indirect assertion calls are
+# assertions that are called within a method that is called by the test being
+# tested.
+#
+# test '1' do
+#   assert true # direct assertion call
+# end
+#
+# test '2' do
+#   validate
+# end
+#
+# def validate
+#   assert true # Direct assertion call if we are unit testing the validate
+#               # method.
+#               # Indirect assertion call if we are unit testing test '2'.
+# end
 class AssertionReporter
   class << self
     def report
       create_csv
-      print_unknown_call_sites
     end
 
     def create_csv
       CSV.open(File.join(__dir__, '..', '..', 'sequence_coverage.csv'), 'wb') do |csv|
-        csv << ['Assertion Location', 'Pass Count', 'Fail Count']
+        csv << [
+          'Assertion Location',
+          'Direct Pass Count',
+          'Direct Fail Count',
+          'Indirect Pass Count',
+          'Indirect Fail Count'
+        ]
         AssertionTracker.assertion_locations.sort.each do |location|
-          results = AssertionTracker.assertion_calls[location]
-          csv << [location, pass_count(results), fail_count(results)]
+          csv << [
+            location,
+            direct_pass_count(location),
+            direct_fail_count(location),
+            indirect_pass_count(location),
+            indirect_fail_count(location)
+          ]
         end
       end
     end
 
-    def pass_count(results)
-      results.count { |result| result }
+    def direct_call?(location)
+      location.include? AssertionCallLocationFormatter::CALL_SEPARATOR
     end
 
-    def fail_count(results)
-      results.count(&:!)
+    def direct_pass_count(location)
+      AssertionTracker.assertion_calls[location].count { |result| result }
```

Comment:
Could also use `.count(&:itself)`.  Not necessarily suggesting it, but it might fit well with how you are counting fails `.count(&:!)` already.

Reply:
Nice, didn't know about that one.